pvlib/iam.py

```diff
@@ -527,3 +528,221 @@ def sapm(aoi, module, upper=None):
         iam = pd.Series(iam, aoi.index)
 
     return iam
+
+
+def marion_diffuse(model, surface_tilt, regions=None, **kwargs):
+    """
+    Determine diffuse irradiance incidence angle modifiers using Marion's
+    method of integrating over solid angle.
+
+    Parameters
+    ----------
+    model : str
+        The IAM function to evaluate across solid angle.  Must be one of
+        `'ashrae', 'physical', 'martin_ruiz', 'sapm'`.
+
+    surface_tilt : numeric
+        Surface tilt angles in decimal degrees.
+        The tilt angle is defined as degrees from horizontal
+        (e.g. surface facing up = 0, surface facing horizon = 90).
+
+    regions : list, default ['sky', 'horizon', 'ground']
+        The regions to integrate over.  Options are:
+
+            * 'sky': radiation from the sky dome (zenith <= 90)
+            * 'horizon': radiation from the region of the sky near the horizon
+              (89.5 <= zenith <= 90)
+            * 'ground': radiation reflected from the ground (zenith >= 90)
+
+        See [1]_ for a detailed description of each class.  If not specified,
+        IAM values for all three regions are calculated.
+
+    **kwargs
+        Extra parameters passed to the IAM function.
+
+    Returns
+    -------
+    iam : dict
+        IAM values for each type of diffuse irradiance.
+
+    See Also
+    --------
+    pvlib.iam.marion_integrate
+
+    References
+    ----------
+    .. [1] B. Marion "Numerical method for angle-of-incidence correction
+       factors for diffuse radiation incident photovoltaic modules",
+       Solar Energy, Volume 147, Pages 344-348. 2017.
+       DOI: 10.1016/j.solener.2017.03.027
+
+    Examples
+    --------
+    >>> marion_diffuse('physical', surface_tilt=20)
+    {'sky': 0.9539178294437575,
+     'horizon': 0.7652650139134007,
+     'ground': 0.6387140117795903}
+
+    >>> marion_diffuse('ashrae', [20, 30], b=0.04)
+    {'sky': array([0.96748999, 0.96938408]),
+     'horizon': array([0.86478428, 0.91825792]),
+     'ground': array([0.77004435, 0.8522436 ])}
+    """
+
+    if regions is None:
+        regions = ['sky', 'horizon', 'ground']
+
+    models = {
+        'physical': physical,
+        'ashrae': ashrae,
+        'sapm': sapm,
+        'martin_ruiz': martin_ruiz,
+    }
+
+    try:
+        iam_model = models[model]
+        iam_function = functools.partial(iam_model, **kwargs)
+    except KeyError:
+        raise ValueError('model must be one of: ' + str(list(models.keys())))
+
+    iam = {}
+    for region in regions:
+        iam[region] = marion_integrate(iam_function, surface_tilt, region)
+
+    return iam
+
+
+def marion_integrate(function, surface_tilt, region, N=None):
```

Comment:
```suggestion
def marion_integrate(function, surface_tilt, region, num=None):
```

Reply:
`num` is consistent with `np.linspace`. I'd also be good with something like `intpnts`.